ports/atmel-samd/common-hal/pulseio/PWMOut.c

```diff
@@ -101,80 +107,91 @@ void common_hal_pulseio_pwmout_construct(pulseio_pwmout_obj_t* self,
-        t = &pin->primary_timer;
-        index = primary_timer_index;
-        self->tcc_instance.hw = t->tcc;
-        self->tcc_instance.double_buffering_enabled = true;
-    } else if (!variable_frequency &&
-               secondary_timer_index != 0xff &&
-               target_timer_frequencies[secondary_timer_index] == frequency &&
-               pin->secondary_timer.tcc->CTRLA.bit.ENABLE == 1 &&
-               channel_ok(&pin->secondary_timer, secondary_timer_index)) {
-        t = &pin->secondary_timer;
-        index = secondary_timer_index;
-        self->tcc_instance.hw = t->tcc;
-        self->tcc_instance.double_buffering_enabled = true;
-    } else {
-        // Pick an unused timer if available.
-
-        // Check the secondary timer first since its always a nicer TCC (when it
-        // exists)
-        if (pin->secondary_timer.tc != 0 &&
-            timer_refcount[secondary_timer_index] == 0 &&
-            pin->secondary_timer.tcc->CTRLA.bit.ENABLE == 0) {
-            t = &pin->secondary_timer;
-            index = secondary_timer_index;
-        } else if (pin->primary_timer.tc != 0 &&
-                   (!pin->primary_timer.is_tc || pin->primary_timer.channel == 1) &&
-                   timer_refcount[primary_timer_index] == 0) {
-            t = &pin->primary_timer;
-            index = primary_timer_index;
+    // First see if a tcc is already going with the frequency we want and our
+    // channel is unused. tc's don't have neough channels to share.
+    const pin_timer_t* timer = NULL;
+    uint8_t mux_position = 0;
+    if (!variable_frequency) {
+        for (uint8_t i = 0; i < TCC_INST_NUM && timer == NULL; i++) {
+            if (target_tcc_frequencies[i] != frequency) {
+                continue;
+            }
+            for (uint8_t j = 0; j < NUM_TIMERS_PER_PIN && timer == NULL; j++) {
+                const pin_timer_t* t = &pin->timer[j];
+                if (t->index != i || t->is_tc || t->index >= TCC_INST_NUM) {
+                    continue;
+                }
+                Tcc* tcc = tcc_insts[t->index];
+                if (tcc->CTRLA.bit.ENABLE == 1 && channel_ok(t)) {
+                    timer = t;
+                    mux_position = j;
+                }
+            }
         }
-        if (t == NULL) {
+    }
+
+    // No existing timer has been found, so find a new one to use and set it up.
+    if (timer == NULL) {
+        // By default, with fixed frequency we want to share a TCC because its likely we'll have
+        // other outputs at the same frequency. If the frequency is variable then we'll only have
+        // one output so we start with the TCs to see if they work.
+        int8_t direction = -1;
+        uint8_t start = NUM_TIMERS_PER_PIN - 1;
+        if (variable_frequency) {
+            direction = 1;
+            start = 0;
+        }
+        for (uint8_t i = start; i >= 0 && i < NUM_TIMERS_PER_PIN && timer == NULL; i += direction) {
```

Comment:
`i` is a `uint8_t`, so it can't be < 0, so not sure why you're checking this. What situation did you have in mind?

Reply:
I hadn't thought about wrap around. I was checking both "ends" because we don't know which direction the loop is running.